### ISSUE TITLE
Z sometimes wrong when simulating Arcs

### ISSUE THREAD

**Issue by Kragrathea** [Kragrathea/OctoPrint-PrettyGCode#86]
When simulating a print the animation for Arcs sometimes uses the wrong Z height.

**Comment by FormerLurker** [comment]
I will see if I can recreate this with simple gcode, and will post it once that is complete.

**Comment by Kragrathea** [comment]
I took a quick look but couldn't immediately see the problem.

As far as debugging it goes. 
=Z shouldn't change during the course of an arc. It should only change with at new layer. And if Z isn't provided as part of the G command it should be assumed to be the same as the last Z. 
-You could try checking the Z value against parserLastExtrudedZ which I use to detect layer changes.

**Comment by FormerLurker** [comment]
I suspect the way I extract z is having problems.  It's different than the model builder code which has methods for ensuring the coordinates are absolute.  If I had to guess what's wrong, that's where I would look.

**Comment by FormerLurker** [comment]
So, this may be solved.  I don't see the old behavior anymore.  The issue boiled down to values being reported as NaN, but that wasn't handled.

I thing some adjustments could be made to the animation routine still, but will have to see how it works on a live print.

**Comment by Kragrathea** [comment]
Cool! Now that you mention it, I think I had the same issue. I see I added IsNan() around everything. I wish I could remember _why_ the values were NaN. Do you happen to know?

**Comment by FormerLurker** [comment]
I do not.  I didn't look into it much, and noticed you were catching those as well.  I did create a function you could use so that they all use the same method (check for undefined and nan).  I could look into this friday though.  I'd like to see if I can figure out the animation bit.

**Comment by Kragrathea** [comment]
This fix works for the Z issue. 

I am having performance issues, but I don't think it is related to this. It might be related to the buffer overflowing. I am looking into that now and will make a separate issue if needed.

**Comment by FormerLurker** [comment]
I had some large memory usage, but chalked it up to viewing a 150mb file..  will take a look and comment in any new issues I see pop up.

**Comment by Kragrathea** [comment]
The way the adaptive buffer works is it speeds up the printhead the longer the buffer is. To overflow you would have to add hundreds of segments at once. Which may be actually be the correct behavior and I just need to make the buffer bigger.

`rate=rate*(buffer.length/5.0);`

**Comment by FormerLurker** [comment]
Maybe an self growing buffer is the way to go?  you could create a function to add to the buffer, and if it hits max_len, grow it by 2x:

```
self.addToBuffer = function(x) {
     if (buffer.length == max_length)
     {
          max_length = max_length * 2;
          buffer.length = max_length * 2;
     }
     buffer.push(x);
}
```

I literally just typed that in, so some debugging/syntax adjustments might be required.

**Comment by Kragrathea** [comment]
This isn't a buffer we want to grow. If it overflows the print head animation just glitches a bit. 

I am going to start a new issue for the buffer overflow.
